gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,179 @@
+import com.google.api.client.auth.oauth2.TokenRequest;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageOptions;
+import com.google.cloud.hadoop.util.AccessTokenProvider;
+import com.google.cloud.hadoop.util.HttpTransportFactory;
+import com.google.cloud.hadoop.util.RedactedString;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Optional;
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ * Retrieve an access token using the OAuth2 refresh token grant flow. See <a
+ * href="https://datatracker.ietf.org/doc/html/rfc6749#section-1.5">RFC 6749</a>.
+ */
+public class RefreshTokenAuth2Provider implements AccessTokenProvider {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+  private static final DateTimeFormatter dateFormat =
+      DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS").withZone(ZoneId.systemDefault());
+  private static final AccessToken EXPIRED_TOKEN = new AccessToken("", -1L);
+
+  private Configuration config;
+  private Optional<AccessToken> accessToken = Optional.empty();
+  private HttpTransport httpTransport;
+  private Optional<RedactedString> previousRefreshToken = Optional.empty();
+
+  @Override
+  public AccessToken getAccessToken() {
+    if (!accessToken.isPresent()) {
+      refresh();
+    }
+    return accessToken.get();
+  }
+
+  @Override
+  public void refresh() {
+    logger.atFine().log(
+        "Refreshing access-token based token. Our token is set to expire at '%s' and it is now '%s'",
+        lazy(
+            () ->
+                dateFormat.format(
+                    Instant.ofEpochMilli(
+                        accessToken.orElse(EXPIRED_TOKEN).getExpirationTimeMilliSeconds()))),
+        lazy(() -> dateFormat.format(Instant.now())));
+
+    GoogleCloudStorageOptions gcsOptions =
+        GoogleHadoopFileSystemConfiguration.getGcsOptionsBuilder(config).build();
+    String tokenServerUrl = gcsOptions.getTokenServerUrl();
+    RedactedString refreshToken = gcsOptions.getRefreshToken();
+    String clientId = gcsOptions.getClientId();
+    RedactedString clientSecret = gcsOptions.getClientSecret();
+    checkNotNull(refreshToken, "Must provide a refresh token");
+    checkNotNull(clientSecret, "Must provide a client secret");
+
+    logger.atFine().log(
+        "Refresh token calling endpoint '%s' with client id '%s'", tokenServerUrl, clientId);
+
+    try {
+      HttpTransport httpTransport = getTransport();
```

Comment:
Should it be class field, so we do not need to initialize HTTP transport for each refresh request?

Reply:
Have a look at the function `getTransport`, it's actually a class field initialised on the fly